File: src/interactive_marker_server.cpp
// Copyright (c) 2011, Willow Garage, Inc.
// All rights reserved.
//
// Software License Agreement (BSD License 2.0)
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
//  * Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
//  * Redistributions in binary form must reproduce the above
//    copyright notice, this list of conditions and the following
//    disclaimer in the documentation and/or other materials provided
//    with the distribution.
//  * Neither the name of Willow Garage, Inc. nor the names of its
//    contributors may be used to endorse or promote products derived
//    from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

// Author: David Gossow

#include <memory>
#include <string>
#include <unordered_map>
#include <utility>

#include "interactive_markers/interactive_marker_server.hpp"

#include "rmw/rmw.h"
#include "rclcpp/qos.hpp"

using visualization_msgs::msg::InteractiveMarkerFeedback;
using visualization_msgs::msg::InteractiveMarkerUpdate;

namespace interactive_markers
{

InteractiveMarkerServer::InteractiveMarkerServer(
  const std::string & topic_ns,
  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr base_interface,
  rclcpp::node_interfaces::NodeClockInterface::SharedPtr clock_interface,
  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr logging_interface,
  rclcpp::node_interfaces::NodeTimersInterface::SharedPtr timers_interface,
  rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr topics_interface,
  rclcpp::node_interfaces::NodeServicesInterface::SharedPtr services_interface,
  const std::string server_id)
: topic_ns_(topic_ns),
  seq_num_(0),
  context_(base_interface->get_context()),
  clock_(clock_interface->get_clock()),
  logger_(logging_interface->get_logger().get_child(server_id))
{
  if (!server_id.empty()) {
    server_id_ = std::string(base_interface->get_fully_qualified_name()) + "/" + server_id;
  } else {
    server_id_ = base_interface->get_fully_qualified_name();
  }

  const std::string update_topic = topic_ns + "/update";
  const std::string feedback_topic = topic_ns + "/feedback";

  get_interactive_markers_service_ = rclcpp::create_service<
    visualization_msgs::srv::GetInteractiveMarkers>(
    base_interface,
    services_interface,
    topic_ns + "/get_interactive_markers",
    std::bind(
      &InteractiveMarkerServer::getInteractiveMarkersCallback,
      this,
      std::placeholders::_1,
      std::placeholders::_2,
      std::placeholders::_3),
    rmw_qos_profile_services_default,
    base_interface->get_default_callback_group());

  rclcpp::QoS update_qos(rclcpp::KeepLast(100));
  update_qos.durability_volatile();
  update_pub_ = rclcpp::create_publisher<InteractiveMarkerUpdate>(topics_interface, update_topic,
      update_qos);

  rclcpp::QoS feedback_qos(rclcpp::KeepLast(100));
  feedback_qos.durability_volatile();
  feedback_sub_ = rclcpp::create_subscription<InteractiveMarkerFeedback>(topics_interface,
      feedback_topic,
      feedback_qos,
      std::bind(&InteractiveMarkerServer::processFeedback, this, std::placeholders::_1));

  keep_alive_timer_ = rclcpp::GenericTimer<std::function<void()>>::make_shared(
    clock_,
    std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::milliseconds(500)),
    std::bind(&InteractiveMarkerServer::keepAlive, this),
    context_);

  rclcpp::callback_group::CallbackGroup::SharedPtr group{nullptr};
  timers_interface->add_timer(keep_alive_timer_, group);
}

InteractiveMarkerServer::~InteractiveMarkerServer()
{
  if (rclcpp::ok(context_)) {
    clear();
    applyChanges();
  }
}

void InteractiveMarkerServer::applyChanges()
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  if (pending_updates_.empty()) {
    return;
  }

  visualization_msgs::msg::InteractiveMarkerUpdate update;
  update.type = visualization_msgs::msg::InteractiveMarkerUpdate::UPDATE;

  update.markers.reserve(marker_contexts_.size());
  update.poses.reserve(marker_contexts_.size());
  update.erases.reserve(marker_contexts_.size());

  for (auto update_it = pending_updates_.begin();
    update_it != pending_updates_.end();
    update_it++)
  {
    M_MarkerContext::iterator marker_context_it = marker_contexts_.find(update_it->first);

    switch (update_it->second.update_type) {
      case UpdateContext::FULL_UPDATE:
        {
          if (marker_context_it == marker_contexts_.end()) {
            RCLCPP_DEBUG(logger_, "Creating new context for %s", update_it->first.c_str());
            // create a new int_marker context
            marker_context_it =
              marker_contexts_.insert(std::make_pair(update_it->first, MarkerContext())).first;
            // copy feedback cbs, in case they have been set before the marker context was created
            marker_context_it->second.default_feedback_cb = update_it->second.default_feedback_cb;
            marker_context_it->second.feedback_cbs = update_it->second.feedback_cbs;
          }

          marker_context_it->second.int_marker = update_it->second.int_marker;

          update.markers.push_back(marker_context_it->second.int_marker);
          break;
        }

      case UpdateContext::POSE_UPDATE:
        {
          if (marker_context_it == marker_contexts_.end()) {
            RCLCPP_ERROR(
              logger_,
              "Pending pose update for non-existing marker found. This is a bug in "
              "InteractiveMarkerInterface.");
          } else {
            marker_context_it->second.int_marker.pose = update_it->second.int_marker.pose;
            marker_context_it->second.int_marker.header = update_it->second.int_marker.header;

            visualization_msgs::msg::InteractiveMarkerPose pose_update;
            pose_update.header = marker_context_it->second.int_marker.header;
            pose_update.pose = marker_context_it->second.int_marker.pose;
            pose_update.name = marker_context_it->second.int_marker.name;
            update.poses.push_back(pose_update);
          }
          break;
        }

      case UpdateContext::ERASE:
        {
          if (marker_context_it != marker_contexts_.end()) {
            marker_contexts_.erase(update_it->first);
            update.erases.push_back(update_it->first);
          }
          break;
        }
    }
  }

  seq_num_++;

  publish(update);
  pending_updates_.clear();
}

bool InteractiveMarkerServer::erase(const std::string & name)
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  if (marker_contexts_.end() == marker_contexts_.find(name) &&
    pending_updates_.end() == pending_updates_.find(name))
  {
    return false;
  }
  pending_updates_[name].update_type = UpdateContext::ERASE;
  return true;
}

void InteractiveMarkerServer::clear()
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  // erase all markers
  pending_updates_.clear();
  M_MarkerContext::iterator it;
  for (it = marker_contexts_.begin(); it != marker_contexts_.end(); it++) {
    pending_updates_[it->first].update_type = UpdateContext::ERASE;
  }
}

bool InteractiveMarkerServer::empty() const
{
  return marker_contexts_.empty();
}

std::size_t InteractiveMarkerServer::size() const
{
  return marker_contexts_.size();
}

bool InteractiveMarkerServer::setPose(
  const std::string & name,
  const geometry_msgs::msg::Pose & pose,
  const std_msgs::msg::Header & header)
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  M_MarkerContext::iterator marker_context_it = marker_contexts_.find(name);
  M_UpdateContext::iterator update_it = pending_updates_.find(name);

  // if there's no marker and no pending addition for it, we can't update the pose
  if (marker_context_it == marker_contexts_.end() &&
    (update_it == pending_updates_.end() ||
    update_it->second.update_type != UpdateContext::FULL_UPDATE))
  {
    return false;
  }

  // keep the old header
  if (header.frame_id.empty()) {
    if (marker_context_it != marker_contexts_.end()) {
      doSetPose(update_it, name, pose, marker_context_it->second.int_marker.header);
    } else if (update_it != pending_updates_.end()) {
      doSetPose(update_it, name, pose, update_it->second.int_marker.header);
    } else {
      RCLCPP_WARN(logger_, "Marker does not exist and there is no pending creation.");
      return false;
    }
  } else {
    doSetPose(update_it, name, pose, header);
  }
  return true;
}

bool InteractiveMarkerServer::setCallback(
  const std::string & name,
  FeedbackCallback feedback_cb,
  uint8_t feedback_type)
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  M_MarkerContext::iterator marker_context_it = marker_contexts_.find(name);
  M_UpdateContext::iterator update_it = pending_updates_.find(name);

  if (marker_context_it == marker_contexts_.end() && update_it == pending_updates_.end()) {
    return false;
  }

  // we need to overwrite both the callbacks for the actual marker
  // and the update, if there's any
  if (marker_context_it != marker_contexts_.end()) {
    // the marker exists, so we can just overwrite the existing callbacks
    if (feedback_type == DEFAULT_FEEDBACK_CB) {
      marker_context_it->second.default_feedback_cb = feedback_cb;
    } else {
      if (feedback_cb) {
        marker_context_it->second.feedback_cbs[feedback_type] = feedback_cb;
      } else {
        marker_context_it->second.feedback_cbs.erase(feedback_type);
      }
    }
  }

  if (update_it != pending_updates_.end()) {
    if (feedback_type == DEFAULT_FEEDBACK_CB) {
      update_it->second.default_feedback_cb = feedback_cb;
    } else {
      if (feedback_cb) {
        update_it->second.feedback_cbs[feedback_type] = feedback_cb;
      } else {
        update_it->second.feedback_cbs.erase(feedback_type);
      }
    }
  }
  return true;
}

void InteractiveMarkerServer::insert(const visualization_msgs::msg::InteractiveMarker & int_marker)
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  M_UpdateContext::iterator update_it = pending_updates_.find(int_marker.name);
  if (update_it == pending_updates_.end()) {
    update_it = pending_updates_.insert(std::make_pair(int_marker.name, UpdateContext())).first;
  }

  update_it->second.update_type = UpdateContext::FULL_UPDATE;
  update_it->second.int_marker = int_marker;
}

void InteractiveMarkerServer::insert(
  const visualization_msgs::msg::InteractiveMarker & int_marker,
  FeedbackCallback feedback_cb,
  uint8_t feedback_type)
{
  insert(int_marker);

  setCallback(int_marker.name, feedback_cb, feedback_type);
}

bool InteractiveMarkerServer::get(
  std::string name,
  visualization_msgs::msg::InteractiveMarker & int_marker) const
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  M_UpdateContext::const_iterator update_it = pending_updates_.find(name);

  if (update_it == pending_updates_.end()) {
    M_MarkerContext::const_iterator marker_context_it = marker_contexts_.find(name);
    if (marker_context_it == marker_contexts_.end()) {
      return false;
    }

    int_marker = marker_context_it->second.int_marker;
    return true;
  }

  // if there's an update pending, we'll have to account for that
  switch (update_it->second.update_type) {
    case UpdateContext::ERASE:
      return false;

    case UpdateContext::POSE_UPDATE:
      {
        M_MarkerContext::const_iterator marker_context_it = marker_contexts_.find(name);
        if (marker_context_it == marker_contexts_.end()) {
          return false;
        }
        int_marker = marker_context_it->second.int_marker;
        int_marker.pose = update_it->second.int_marker.pose;
        return true;
      }

    case UpdateContext::FULL_UPDATE:
      int_marker = update_it->second.int_marker;
      return true;
  }

  return false;
}

void InteractiveMarkerServer::getInteractiveMarkersCallback(
  const std::shared_ptr<rmw_request_id_t> request_header,
  const std::shared_ptr<visualization_msgs::srv::GetInteractiveMarkers::Request> request,
  std::shared_ptr<visualization_msgs::srv::GetInteractiveMarkers::Response> response)
{
  (void)request_header;

  RCLCPP_DEBUG(logger_, "Responding to request to get interactive markers");
  response->server_id = server_id_;
  response->sequence_number = seq_num_;
  response->markers.reserve(marker_contexts_.size());
  M_MarkerContext::iterator it;
  for (it = marker_contexts_.begin(); it != marker_contexts_.end(); it++) {
    RCLCPP_DEBUG(logger_, "Sending marker '%s'", it->second.int_marker.name.c_str());
    response->markers.push_back(it->second.int_marker);
  }
}

void InteractiveMarkerServer::processFeedback(
  visualization_msgs::msg::InteractiveMarkerFeedback::ConstSharedPtr feedback)
{
  std::unique_lock<std::recursive_mutex> lock(mutex_);

  M_MarkerContext::iterator marker_context_it = marker_contexts_.find(feedback->marker_name);

  // ignore feedback for non-existing markers
  if (marker_context_it == marker_contexts_.end()) {
    return;
  }

  MarkerContext & marker_context = marker_context_it->second;

  // if two callers try to modify the same marker, reject (timeout= 1 sec)
  if (marker_context.last_client_id != feedback->client_id &&
    (clock_->now() - marker_context.last_feedback).seconds() < 1.0)
  {
    RCLCPP_DEBUG(logger_, "Rejecting feedback for %s: conflicting feedback from separate clients.",
      feedback->marker_name.c_str());
    return;
  }

  marker_context.last_feedback = clock_->now();
  marker_context.last_client_id = feedback->client_id;

  if (feedback->event_type == visualization_msgs::msg::InteractiveMarkerFeedback::POSE_UPDATE) {
    if (marker_context.int_marker.header.stamp == rclcpp::Time()) {
      // keep the old header
      doSetPose(pending_updates_.find(
          feedback->marker_name), feedback->marker_name, feedback->pose,
        marker_context.int_marker.header);
    } else {
      doSetPose(pending_updates_.find(
          feedback->marker_name), feedback->marker_name, feedback->pose, feedback->header);
    }
  }

  // call feedback handler
  std::unordered_map<uint8_t, FeedbackCallback>::iterator feedback_cb_it =
    marker_context.feedback_cbs.find(feedback->event_type);
  if (feedback_cb_it != marker_context.feedback_cbs.end() && feedback_cb_it->second) {
    // call type-specific callback
    feedback_cb_it->second(feedback);
  } else if (marker_context.default_feedback_cb) {
    // call default callback
    marker_context.default_feedback_cb(feedback);
  }
}

void InteractiveMarkerServer::keepAlive()
{
  visualization_msgs::msg::InteractiveMarkerUpdate empty_update;
  empty_update.type = visualization_msgs::msg::InteractiveMarkerUpdate::KEEP_ALIVE;
  publish(empty_update);
}

void InteractiveMarkerServer::publish(visualization_msgs::msg::InteractiveMarkerUpdate & update)
{
  update.server_id = server_id_;
  update.seq_num = seq_num_;
  update_pub_->publish(update);
}

void InteractiveMarkerServer::doSetPose(
  M_UpdateContext::iterator update_it,
  const std::string & name,
  const geometry_msgs::msg::Pose & pose,
  const std_msgs::msg::Header & header)
{
  if (update_it == pending_updates_.end()) {
    update_it = pending_updates_.insert(std::make_pair(name, UpdateContext())).first;
    update_it->second.update_type = UpdateContext::POSE_UPDATE;
  } else if (update_it->second.update_type != UpdateContext::FULL_UPDATE) {
    update_it->second.update_type = UpdateContext::POSE_UPDATE;
  }

  update_it->second.int_marker.pose = pose;
  update_it->second.int_marker.header = header;
  RCLCPP_DEBUG(logger_, "Marker '%s' is now at %f, %f, %f",
    update_it->first.c_str(), pose.position.x, pose.position.y, pose.position.z);
}

}  // namespace interactive_markers

File: interactive_markers/interactive_marker_server.py
# Copyright (c) 2011, Willow Garage, Inc.
# All rights reserved.
#
# Software License Agreement (BSD License 2.0)
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Willow Garage, Inc. nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# Author: Michael Ferguson

import rclpy
from rclpy.duration import Duration
from rclpy.time import Time

from std_msgs.msg import Header

from visualization_msgs.msg import InteractiveMarker
from visualization_msgs.msg import InteractiveMarkerFeedback
from visualization_msgs.msg import InteractiveMarkerPose
from visualization_msgs.msg import InteractiveMarkerUpdate
from visualization_msgs.srv import GetInteractiveMarkers

from threading import Lock


class MarkerContext:
    """Represents a single marker."""

    def __init__(self, time):
        self.last_feedback = time
        self.last_client_id = ""
        self.default_feedback_callback = None
        self.feedback_callbacks = dict()
        self.int_marker = InteractiveMarker()


class UpdateContext:
    """Represents an update to a single marker."""

    FULL_UPDATE = 0
    POSE_UPDATE = 1
    ERASE = 2

    def __init__(self):
        self.update_type = self.FULL_UPDATE
        self.int_marker = InteractiveMarker()
        self.default_feedback_callback = None
        self.feedback_callbacks = dict()


class InteractiveMarkerServer:
    """
    A server to one or many clients (e.g. rviz) displaying a set of interactive markers.

    Note: Keep in mind that changes made by calling insert(), erase(), setCallback() etc.
          are not applied until calling applyChanges().
    """

    DEFAULT_FEEDBACK_CALLBACK = 255

    def __init__(self, node, topic_ns, q_size=100):
        """
        Create an InteractiveMarkerServer and associated ROS connections.

        :param node: The node to attach this interactive marker server to.
        :param topic_ns: The interface will use the topics topic_ns/update and topic_ns/feedback
            for communication.
        """
        self.node = node
        self.topic_ns = topic_ns
        self.seq_num = 0
        self.mutex = Lock()

        # contains the current state of all markers
        # string : MarkerContext
        self.marker_contexts = dict()

        # updates that have to be sent on the next publish
        # string : UpdateContext
        self.pending_updates = dict()

        self.get_interactive_markers_srv = self.node.create_service(
            GetInteractiveMarkers,
            topic_ns + "/get_interactive_markers",
            self.getInteractiveMarkersCallback
        )

        self.update_pub = self.node.create_publisher(InteractiveMarkerUpdate, topic_ns + "/update", q_size)

        self.feedback_sub = self.node.create_subscription(InteractiveMarkerFeedback, topic_ns + "/feedback", self.processFeedback, q_size)

    def shutdown(self):
        """
        Shutdown the interactive marker server.

        This should be called before the node is destroyed so that the internal ROS entities
        can be destroyed.
        """
        self.clear()
        self.applyChanges()
        self.get_interactive_markers_srv = None
        self.update_pub = None
        self.feedback_sub = None

    def __del__(self):
        """Destruction of the interface will lead to all managed markers being cleared."""

        self.shutdown()

    def insert(self, marker, *, feedback_callback=None, feedback_type=DEFAULT_FEEDBACK_CALLBACK):
        """
        Add or replace a marker.

        Note: Changes to the marker will not take effect until you call applyChanges().
        The callback changes immediately.

        :param marker: The marker to be added or replaced.
        :param feedback_callback: Function to call on the arrival of a feedback message.
        :param feedback_type: Type of feedback for which to call the feedback.
        """

        with self.mutex:
            try:
                update = self.pending_updates[marker.name]
            except:
                update = UpdateContext()
                self.pending_updates[marker.name] = update
            update.update_type = UpdateContext.FULL_UPDATE
            update.int_marker = marker
        if feedback_callback is not None:
            self.setCallback(marker.name, feedback_callback, feedback_type)

    def setPose(self, name, pose, header=Header()):
        """
        Update the pose of a marker with the specified name.

        Note: This change will not take effect until you call applyChanges()

        :param name: Name of the interactive marker.
        :param pose: The new pose.
        :param header: Header replacement. Leave this empty to use the previous one.
        :return: True if a marker with that name exists, False otherwise.
        """

        with self.mutex:
            try:
                marker_context = self.marker_contexts[name]
            except:
                marker_context = None
            try:
                update = self.pending_updates[name]
            except:
                update = None
            # if there's no marker and no pending addition for it, we can't update the pose
            if marker_context is None and update is None:
                return False
            if update is not None and update.update_type == UpdateContext.FULL_UPDATE:
                return False

            if header.frame_id is None or header.frame_id == "":
                # keep the old header
                self.doSetPose(update, name, pose, marker_context.int_marker.header)
            else:
                self.doSetPose(update, name, pose, header)
            return True

    def erase(self, name):
        """
        Erase the marker with the specified name.

        Note: This change will not take effect until you call applyChanges().

        :param name: Name of the interactive marker.
        :return: True if a marker with that name exists, False otherwise.
        """

        with self.mutex:
            try:
                self.pending_updates[name].update_type = UpdateContext.ERASE
                return True
            except:
                try:
                    self.marker_contexts[name]  # check exists
                    update = UpdateContext()
                    update.update_type = UpdateContext.ERASE
                    self.pending_updates[name] = update
                    return True
                except:
                    return False

    def clear(self):
        """
        Clear all markers.

        Note: This change will not take effect until you call applyChanges().
        """

        self.pending_updates = dict()
        for marker_name in self.marker_contexts.keys():
            self.erase(marker_name)

    def setCallback(self, name, feedback_callback, feedback_type=DEFAULT_FEEDBACK_CALLBACK):
        """
        Add or replace a callback function for the specified marker.

        Note: This change will not take effect until you call applyChanges().
        The server will try to call any type-specific callback first.
        If a callback for the given type already exists, it will be replaced.
        To unset a callback, pass a value of None.

        :param name: Name of the interactive marker
        :param feedback_callback: Function to call on the arrival of a feedback message.
        :param feedback_type: Type of feedback for which to call the feedback.
            Leave this empty to make this the default callback.
        """

        with self.mutex:
            try:
                marker_context = self.marker_contexts[name]
            except:
                marker_context = None
            try:
                update = self.pending_updates[name]
            except:
                update = None
            if marker_context is None and update is None:
                return False

            # we need to overwrite both the callbacks for the actual marker
            # and the update, if there's any
            if marker_context:
                # the marker exists, so we can just overwrite the existing callbacks
                if feedback_type == self.DEFAULT_FEEDBACK_CALLBACK:
                    marker_context.default_feedback_callback = feedback_callback
                else:
                    if feedback_callback:
                        marker_context.feedback_callbacks[feedback_type] = feedback_callback
                    elif feedback_type in marker_context.feedback_callbacks:
                        del marker_context.feedback_callbacks[feedback_type]
            if update:
                if feedback_type == self.DEFAULT_FEEDBACK_CALLBACK:
                    update.default_feedback_callback = feedback_callback
                else:
                    if feedback_callback:
                        update.feedback_callbacks[feedback_type] = feedback_callback
                    elif feedback_type in update.feedback_callbacks:
                        del update.feedback_callbacks[feedback_type]
            return True

    def applyChanges(self):
        """Apply changes made since the last call to this method and broadcast to clients."""

        with self.mutex:
            if len(self.pending_updates.keys()) == 0:
                return

            update_msg = InteractiveMarkerUpdate()
            update_msg.type = InteractiveMarkerUpdate.UPDATE

            for name, update in self.pending_updates.items():
                if update.update_type == UpdateContext.FULL_UPDATE:
                    try:
                        marker_context = self.marker_contexts[name]
                    except:
                        self.node.get_logger().debug("Creating new context for " + name)
                        # create a new int_marker context
                        marker_context = MarkerContext(self.node.get_clock().now())
                        marker_context.default_feedback_callback = update.default_feedback_callback
                        marker_context.feedback_callbacks = update.feedback_callbacks
                        self.marker_contexts[name] = marker_context

                    marker_context.int_marker = update.int_marker
                    update_msg.markers.append(marker_context.int_marker)

                elif update.update_type == UpdateContext.POSE_UPDATE:
                    try:
                        marker_context = self.marker_contexts[name]
                        marker_context.int_marker.pose = update.int_marker.pose
                        marker_context.int_marker.header = update.int_marker.header

                        pose_update = InteractiveMarkerPose()
                        pose_update.header = marker_context.int_marker.header
                        pose_update.pose = marker_context.int_marker.pose
                        pose_update.name = marker_context.int_marker.name
                        update_msg.poses.append(pose_update)
                    except:
                        self.node.get_logger().error(
                            'Pending pose update for non-existing marker found. '
                            'This is a bug in InteractiveMarkerServer.')

                elif update.update_type == UpdateContext.ERASE:
                    try:
                        marker_context = self.marker_contexts[name]
                        del self.marker_contexts[name]
                        update_msg.erases.append(name)
                    except:
                        pass
                self.pending_updates = dict()

        self.seq_num += 1
        self.publish(update_msg)

    def get(self, name):
        """
        Get marker by name.

        :param name: Name of the interactive marker.
        :return: Marker if exists, None otherwise.
        """

        try:
            update = self.pending_updates[name]
        except:
            try:
                marker_context = self.marker_contexts[name]
            except:
                return None
            return marker_context.int_marker

        # if there's an update pending, we'll have to account for that
        if update.update_type == UpdateContext.ERASE:
            return None
        elif update.update_type == UpdateContext.POSE_UPDATE:
            try:
                marker_context = self.marker_contexts[name]
            except:
                return None
            int_marker = marker_context.int_marker
            int_marker.pose = update.int_marker.pose
            return int_marker
        elif update.update_type == UpdateContext.FULL_UPDATE:
            return update.int_marker
        return None

    def processFeedback(self, feedback):
        """Update marker pose and call user callback."""

        with self.mutex:
            # ignore feedback for non-existing markers
            try:
                marker_context = self.marker_contexts[feedback.marker_name]
            except:
                return

            # if two callers try to modify the same marker, reject (timeout= 1 sec)
            if marker_context.last_client_id != feedback.client_id \
               and (self.node.get_clock().now() - marker_context.last_feedback) < Duration(seconds=1.0):
                self.node.get_logger().debug("Rejecting feedback for " +
                               feedback.marker_name +
                               ": conflicting feedback from separate clients.")
                return

            marker_context.last_feedback = self.node.get_clock().now()
            marker_context.last_client_id = feedback.client_id

            if feedback.event_type == feedback.POSE_UPDATE:
                if marker_context.int_marker.header.stamp == Time(clock_type=self.node.get_clock().clock_type):
                    # keep the old header
                    try:
                        self.doSetPose(self.pending_updates[feedback.marker_name],
                                       feedback.marker_name,
                                       feedback.pose,
                                       marker_context.int_marker.header)
                    except:
                        self.doSetPose(None,
                                       feedback.marker_name,
                                       feedback.pose,
                                       marker_context.int_marker.header)
                else:
                    try:
                        self.doSetPose(self.pending_updates[feedback.marker_name],
                                       feedback.marker_name,
                                       feedback.pose,
                                       feedback.header)
                    except:
                        self.doSetPose(None, feedback.marker_name, feedback.pose, feedback.header)

        # call feedback handler
        feedback_callback = marker_context.feedback_callbacks.get(
            feedback.event_type, marker_context.default_feedback_callback)
        if feedback_callback is not None:
            feedback_callback(feedback)

    def publish(self, update):
        """Increase the sequence number and publish an update."""

        update.seq_num = self.seq_num
        self.update_pub.publish(update)

    def getInteractiveMarkersCallback(self, request, response):
        """Process a service request to get the current interactive markers."""

        with self.mutex:
            response.sequence_number = self.seq_num

            self.node.get_logger().debug('Markers requested. Responding with the following markers:')
            for name, marker_context in self.marker_contexts.items():
                self.node.get_logger().debug('    ' + name)
                response.markers.append(marker_context.int_marker)

            return response

    def doSetPose(self, update, name, pose, header):
        """Schedule a pose update pose without locking."""

        if update is None:
            update = UpdateContext()
            update.update_type = UpdateContext.POSE_UPDATE
            self.pending_updates[name] = update
        elif update.update_type != UpdateContext.FULL_UPDATE:
            update.update_type = UpdateContext.POSE_UPDATE

        update.int_marker.pose = pose
        update.int_marker.header = header
        self.node.get_logger().debug("Marker '" + name + "' is now at " +
                       str(pose.position.x) + ", " + str(pose.position.y) +
                       ", " + str(pose.position.z))
